File: scripts/fetch-toilets.py
"""
Overpass APIから東京・大阪・京都のトイレデータを取得してJSONに保存
"""
import json
import os
import time
import urllib.request
import urllib.parse

CITIES = {
    "tokyo": "35.5,139.4,35.9,139.9",
    "osaka": "34.5,135.3,34.8,135.7",
    "kyoto": "34.9,135.6,35.1,135.9",
}

BLACKLIST_AMENITIES = {"bar", "nightclub", "pub", "casino", "stripclub", "amusement_arcade", "gambling"}
BLACKLIST_KEYWORDS = ["パチンコ", "スロット", "カラオケ", "バー", "スナック", "ナイトクラブ", "キャバクラ", "風俗"]


def is_whitelisted(tags):
    if tags.get("access") == "private":
        return False
    if tags.get("amenity") in BLACKLIST_AMENITIES:
        return False
    if tags.get("leisure") == "amusement_arcade":
        return False
    operator = (tags.get("operator", "") + tags.get("name", "")).lower()
    for kw in BLACKLIST_KEYWORDS:
        if kw.lower() in operator:
            return False
    return True


def fetch_city(city_name, bbox):
    query = f"""
[out:json][timeout:60];
(
  node["amenity"="toilets"]({bbox});
  way["amenity"="toilets"]({bbox});
);
out center tags;
""".strip()

    url = "https://overpass-api.de/api/interpreter"
    data = urllib.parse.urlencode({"data": query}).encode("utf-8")
    req = urllib.request.Request(url, data=data, method="POST")
    req.add_header("Content-Type", "application/x-www-form-urlencoded")
    req.add_header("User-Agent", "FamilyToiletJapan/1.0")

    print(f"Fetching {city_name}...")
    with urllib.request.urlopen(req, timeout=90) as res:
        result = json.loads(res.read().decode("utf-8"))

    elements = result.get("elements", [])
    print(f"  Raw results: {len(elements)}")

    toilets = []
    for el in elements:
        tags = el.get("tags", {})
        if not is_whitelisted(tags):
            continue

        lat = el.get("lat") or (el.get("center") or {}).get("lat")
        lon = el.get("lon") or (el.get("center") or {}).get("lon")
        if not lat or not lon:
            continue

        changing = tags.get("changing_table") == "yes" or tags.get("toilets:changing_table") == "yes"
        toilets.append({
            "id": f"osm-{el['type']}-{el['id']}",
            "lat": lat,
            "lon": lon,
            "name": tags.get("name") or tags.get("name:en"),
            "operator": tags.get("operator"),
            "changingTable": changing,
            "wheelchair": tags.get("wheelchair") == "yes",
            "openingHours": tags.get("opening_hours"),
            "fee": tags.get("fee") == "yes",
            "city": city_name,
            "source": "osm",
        })

    print(f"  After filter: {len(toilets)}")
    return toilets


def main():
    all_toilets = []
    for city_name, bbox in CITIES.items():
        try:
            toilets = fetch_city(city_name, bbox)
            all_toilets.extend(toilets)
            time.sleep(2)
        except Exception as e:
            print(f"  Error: {e}")

    out_dir = os.path.join(os.path.dirname(__file__), "..", "public", "data")
    os.makedirs(out_dir, exist_ok=True)
    out_path = os.path.join(out_dir, "toilets.json")
    with open(out_path, "w", encoding="utf-8") as f:
        json.dump(all_toilets, f, ensure_ascii=False, indent=2)

    print(f"\nTotal: {len(all_toilets)} toilets saved to {out_path}")
    for city in CITIES:
        count = sum(1 for t in all_toilets if t["city"] == city)
        changing = sum(1 for t in all_toilets if t["city"] == city and t["changingTable"])
        print(f"  {city}: {count} total, {changing} with changing table")


if __name__ == "__main__":
    main()

File: scripts/fetch-toilets.ts
/**
 * Overpass APIから東京・大阪・京都のトイレデータを取得してJSONに保存するスクリプト
 * 実行: npx ts-node scripts/fetch-toilets.ts
 */

import * as fs from "fs";
import * as path from "path";

interface OverpassElement {
  type: "node" | "way" | "relation";
  id: number;
  lat?: number;
  lon?: number;
  center?: { lat: number; lon: number };
  tags?: Record<string, string>;
}

interface OverpassResponse {
  elements: OverpassElement[];
}

export interface Toilet {
  id: string;
  lat: number;
  lon: number;
  name?: string;
  operator?: string;
  changingTable: boolean;
  wheelchair: boolean;
  openingHours?: string;
  fee: boolean;
  city: string;
  source: "osm";
}

// ホワイトリスト：表示対象の運営者・施設種別
const WHITELIST_OPERATORS = [
  "公園", "park", "駅", "station", "道の駅", "michi-no-eki",
  "セブン", "7-eleven", "seven", "ローソン", "lawson",
  "ファミリーマート", "familymart", "family mart",
  "ショッピング", "shopping", "mall", "百貨店", "department",
  "博物館", "museum", "美術館", "図書館", "library",
  "病院", "hospital", "クリニック", "clinic",
  "観光", "tourist", "公共", "public", "municipal",
  "東京都", "大阪府", "京都府", "区", "市",
];

// ブラックリスト：除外対象
const BLACKLIST_OPERATORS = [
  "パチンコ", "pachinko", "スロット", "カラオケ", "karaoke",
  "バー", "bar", "スナック", "snack", "ナイトクラブ", "nightclub",
  "キャバクラ", "風俗",
];

const BLACKLIST_AMENITIES = [
  "bar", "nightclub", "pub", "casino", "stripclub",
  "amusement_arcade", "gambling",
];

function isWhitelisted(element: OverpassElement): boolean {
  const tags = element.tags || {};

  // ブラックリストチェック
  if (BLACKLIST_AMENITIES.includes(tags.amenity || "")) return false;
  if (tags.leisure === "amusement_arcade") return false;

  const operator = (tags.operator || tags.name || "").toLowerCase();
  if (BLACKLIST_OPERATORS.some((b) => operator.includes(b.toLowerCase()))) {
    return false;
  }

  // access=privateは除外
  if (tags.access === "private") return false;

  return true;
}

function buildQuery(city: string, bbox: string): string {
  return `
[out:json][timeout:60];
(
  node["amenity"="toilets"](${bbox});
  way["amenity"="toilets"](${bbox});
);
out center tags;
`.trim();
}

const CITIES: Record<string, { bbox: string }> = {
  tokyo: { bbox: "35.5, 139.4, 35.9, 139.9" },
  osaka: { bbox: "34.5, 135.3, 34.8, 135.7" },
  kyoto: { bbox: "34.9, 135.6, 35.1, 135.9" },
};

async function fetchCity(cityName: string, bbox: string): Promise<Toilet[]> {
  const query = buildQuery(cityName, bbox);
  const url = "https://overpass-api.de/api/interpreter";

  console.log(`Fetching ${cityName}...`);

  const res = await fetch(url, {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    body: `data=${encodeURIComponent(query)}`,
  });

  if (!res.ok) {
    throw new Error(`Overpass API error: ${res.status} ${res.statusText}`);
  }

  const data: OverpassResponse = await res.json();
  console.log(`  Raw results: ${data.elements.length}`);

  const toilets: Toilet[] = [];

  for (const el of data.elements) {
    if (!isWhitelisted(el)) continue;

    const lat = el.lat ?? el.center?.lat;
    const lon = el.lon ?? el.center?.lon;
    if (!lat || !lon) continue;

    const tags = el.tags || {};

    toilets.push({
      id: `osm-${el.type}-${el.id}`,
      lat,
      lon,
      name: tags.name || tags["name:en"] || undefined,
      operator: tags.operator || undefined,
      changingTable:
        tags.changing_table === "yes" || tags["toilets:changing_table"] === "yes",
      wheelchair: tags.wheelchair === "yes",
      openingHours: tags.opening_hours || undefined,
      fee: tags.fee === "yes",
      city: cityName,
      source: "osm",
    });
  }

  console.log(`  After filter: ${toilets.length}`);
  return toilets;
}

async function main() {
  const allToilets: Toilet[] = [];

  for (const [cityName, { bbox }] of Object.entries(CITIES)) {
    try {
      const toilets = await fetchCity(cityName, bbox);
      allToilets.push(...toilets);
      // Overpass APIへの負荷を避けるため1秒待機
      await new Promise((r) => setTimeout(r, 1000));
    } catch (err) {
      console.error(`Error fetching ${cityName}:`, err);
    }
  }

  const outDir = path.join(process.cwd(), "public", "data");
  fs.mkdirSync(outDir, { recursive: true });

  const outPath = path.join(outDir, "toilets.json");
  fs.writeFileSync(outPath, JSON.stringify(allToilets, null, 2), "utf-8");

  console.log(`\nTotal: ${allToilets.length} toilets saved to ${outPath}`);

  // 都市別サマリー
  for (const city of Object.keys(CITIES)) {
    const count = allToilets.filter((t) => t.city === city).length;
    const withChanging = allToilets.filter(
      (t) => t.city === city && t.changingTable
    ).length;
    console.log(`  ${city}: ${count} total, ${withChanging} with changing table`);
  }
}

main().catch(console.error);
